### PR TITLE
Resolve #115: Add ContravariantRegistrationSource docs.

### LINKED ISSUE #115: Documentation on ContravariantRegistrationSource
Autofac has a hidden gem `ContravariantRegistrationSource` but there is no documentation on it so folks don't know about it. Some docs with an example explaining usage (and why you'd want to use it) would be great.

### PR DESCRIPTION
**Pull request by tillig**
Resolves #115. Adds docs to the registration source page to discuss the contravariant registration source.

I'll admit sometimes I get confused with covariance vs. contravariance, too, so double-check my example. I tried to simplify the [tests for `ContravariantRegistrationSource`](https://github.com/autofac/Autofac/blob/master/test/Autofac.Test/Features/Variance/ContravariantRegistrationSourceTests.cs) down to something digestible but admittedly didn't run the whole thing through a compiler or anything.

### CODE REVIEW

File: docs/advanced/registration-sources.rst [hunk 1]
```diff
@@ -20,6 +20,42 @@ You can use the ``Autofac.Features.ResolveAnything.AnyConcreteTypeNotAlreadyRegi
...
+    }
+
+    var builder = new ContainerBuilder();
+    builder.RegisterSource(new ContravariantRegistrationSource());
+    builder.RegisterType<CommandAHandler>().As<IHandler<CommandA>>();
+    var container = builder.Build();
+
+    // a and b are both CommandAHandler.
+    var a = container.Resolve<IHandler<CommandA>>();
+    var b = container.Resolve<IHandler<CommandB>>();
```
Comment (alistairjevans) [inline docs/advanced/registration-sources.rst]: Might I propose an example slightly closer to contravariance names, and using a similar Handle method to the example in the linked MS doc?
```suggestion
    public class BaseCommandHandler : IHandler<BaseCommand>
    {
      public void Handle(BaseCommand command)
      {
         Console.WriteLine(command.GetType().Name);
      }
    }

    public class BaseCommand
    {
    }

    public class DerivedCommand : BaseCommand
    {
    }

    var builder = new ContainerBuilder();
    builder.RegisterSource(new ContravariantRegistrationSource());
    builder.RegisterType<BaseCommandHandler>().As<IHandler<BaseCommand>>();
    var container = builder.Build();

    // a and b are both BaseCommandHandler.
    var a = container.Resolve<IHandler<BaseCommand>>();
    a.Handle(new BaseCommand()); // prints BaseCommand
    var b = container.Resolve<IHandler<DerivedCommand>>();
    b.Handle(new DerivedCommand()); // prints DerivedCommand
```

**Review by alistairjevans — CHANGES_REQUESTED** [review]
Proposed a change to the example.

File: docs/advanced/registration-sources.rst (hunk 1 above)
Comment (tillig) [reply docs/advanced/registration-sources.rst]: Sounds good to me. Committed. I wasn't sure how large or small the example should be. Somewhat debated not putting an example and referring people to the unit tests. ;)

Comment (alistairjevans) [reply docs/advanced/registration-sources.rst]: Cruel, but not unusual...

**Pull request merged**